lib/recurly/validate.js

```diff
@@ -117,34 +117,59 @@ function buildCompareValue (start, length, terminator) {
  */
 
 export function cardType (number, partial = false) {
-  const cardNumber = parseCard(number);
-  const compareLength = Math.min(cardNumber.length, 6);
+  return findCardGroup(number, partial)?.type || 'unknown';
+}
 
-  const compareValue = buildCompareValue(cardNumber, compareLength, '0');
+function findCardGroup (number, partial) {
+  const cardNumber = parseCard(number);
 
-  const types = Object.keys(CREDIT_CARD_TYPES).filter((type) => {
+  const types = Object.keys(CREDIT_CARD_TYPES).map(function (type) {
     if (partial && type == 'maestro') {
       // Maestro has a wide range (6*) that overlaps with some other types,
       // which can be disambiguated only when the full lenght is given
       return;
     }
 
-    return find(CREDIT_CARD_TYPES[type], ((group) => {
+    const group = find(CREDIT_CARD_TYPES[type], ((group) => {
       if (!partial && group.lengths.indexOf(cardNumber.length) < 0) {
         return false;
       }
 
-      return find(group.ranges, ([rangeBegin, rangeEnd]) => {
-        const start = buildCompareValue(rangeBegin, compareLength, '0');
-        const end = buildCompareValue(rangeEnd, compareLength, '9');
-
-        return compareValue >= start && compareValue <= end;
-      });
+      return findInRange(group.ranges, cardNumber);
     }));
+
+    if (group) {
+      return { type, group };
+    }
+  }).filter(function (typeWithGroup) { return !!typeWithGroup; });
```

Comment:
This could be reduced from two iterations to one by changing the above `map` to a `filter` and updating the return conditional to:

```
return group ? { type, group } : false;
```

Reply:
👍 sure thing. A reduce would be possible, but it’s no more efficient and less straightforward.

Reply:
👍 sure thing. A reduce would be possible, but it’s no more efficient and less straightforward.